Applications/Spire/Include/Spire/Ui/AnyInputBox.hpp

```diff
@@ -0,0 +1,183 @@
+      template<typename T>
+      explicit AnyInputBox(T& input_box, QWidget* parent = nullptr);
+
+      /** Returns the current value model. */
+      const std::shared_ptr<AnyValueModel>& get_current() const;
+
+      /** Returns the last submission. */
+      const AnyRef& get_submission() const;
+
+      /** Returns <code>true</code> iff this input box is read-only. */
+      bool is_read_only() const;
+
+      /** Sets whether this input box is read-only. */
+      void set_read_only(bool read_only);
+
+      /** Connects a slot to the SubmitSignal. */
+      boost::signals2::connection connect_submit_signal(
+        const SubmitSignal::slot_type& slot) const;
+
+      /** Connects a slot to the RejectedSignal. */
+      boost::signals2::connection connect_reject_signal(
+        const RejectSignal::slot_type& slot) const;
+
+    private:
+      struct AbstractInputBox {
+        virtual ~AbstractInputBox() = default;
+        virtual const std::shared_ptr<AnyValueModel>& get_current() const = 0;
+        virtual const AnyRef& get_submission() = 0;
+        virtual bool is_read_only() const = 0;
+        virtual void set_read_only(bool read_only) = 0;
+        virtual boost::signals2::connection connect_submit_signal(
+          const SubmitSignal::slot_type& slot) const = 0;
+        virtual boost::signals2::connection connect_reject_signal(
+          const RejectSignal::slot_type& slot) const = 0;
+      };
+      template<typename T>
+      struct WrapperInputBox final : AbstractInputBox {
+        mutable SubmitSignal m_submit_signal;
+        mutable RejectSignal m_reject_signal;
+        T* m_input_box;
+        std::shared_ptr<AnyValueModel> m_current;
+        AnyRef m_submission;
+        boost::signals2::scoped_connection m_submit_connection;
+        boost::signals2::scoped_connection m_reject_connection;
+
+        WrapperInputBox(T& input_box);
+        const std::shared_ptr<AnyValueModel>& get_current() const override;
+        const AnyRef& get_submission() override;
+        bool is_read_only() const override;
+        void set_read_only(bool read_only) override;
+        boost::signals2::connection connect_submit_signal(
+          const SubmitSignal::slot_type& slot) const override;
+        boost::signals2::connection connect_reject_signal(
+          const RejectSignal::slot_type& slot) const override;
+      };
+      std::unique_ptr<AbstractInputBox> m_input_box;
+  };
+
+  template<typename T>
+  AnyInputBox::AnyInputBox(T& input_box, QWidget* parent)
+      : QWidget(parent),
+        m_input_box(std::make_unique<WrapperInputBox<T>>(input_box)) {
+    auto layout = new QHBoxLayout(this);
+    layout->setContentsMargins({});
+    layout->addWidget(&input_box);
+  }
+
+  template<typename T>
+  AnyInputBox::WrapperInputBox<T>::WrapperInputBox(T& input_box)
+    : m_input_box(&input_box),
+      m_current(std::make_shared<AnyValueModel>(
+        std::static_pointer_cast<ValueModel<Details::input_current_type_t<T>>>(
```

Comment:
Remove all these unnecessary casts.

Reply:
The `get_current` functions of `DateBox` and `DecimalBox` return respectively `std::shared_ptr<OptionalDateModel>` and `std::shared_ptr<OptionalDecimalModel>` which can't be implicitly casted to `std::shared_ptr<ValueModel<boost::optional<boost::gregorian::date>>>` or `std::shared_ptr<ValueModel<boost::optional<Decimal>>>` in the constructor template of the `AnyValueModel`, so I had to do these explicit casts.

Reply:
Based on the workaround proposed in [Why isn't shared_ptr to Derived implicitly converted to shared_ptr to Base](https://stackoverflow.com/questions/64468726/why-isnt-shared-ptr-to-derived-implicitly-converted-to-shared-ptr-to-base), I propose an update of the AnyValueModel's constructor:

```      
template<typename T,
   std::enable_if_t<std::is_convertible_v<T, std::shared_ptr<ValueModel<typename T::element_type::Type>>>, bool> = true>
explicit AnyValueModel(T source);
```